torchsde/_core/adjoint_sde.py

```diff
@@ -58,32 +58,72 @@ def __init__(self,
             NOISE_TYPES.diagonal: self.gdg_prod_diagonal,
         }.get(sde.noise_type, self.gdg_prod_default)
 
-    def _unpack_y_aug(self, y_aug):
+    def _unpack_y_aug(self, t, y_aug, v):
+        # These leaf checks are very important.
+        # _unpack_y_aug is used where we want to compute:
+        # ```
+        # with torch.enable_grad():
+        #     s = some_function(y)
+        #     torch.autograd.grad(s, [y] + params, ...)
+        # ```
+        # where `some_function` implicitly depends on `params`.
+        # However if y has history of its own then in principle it could _also_ depend upon `params`, and this call to
+        # `grad` will go all the way back to that.
+        # To avoid this, we require that y (and the other tensors) be leaf tensors.
+        #
+        # Note that this is also the reason for the `y0.detach()` in adjoint.py::_SdeintAdjointMethod.forward. Unless
+        # we detach then y0 may have a history and these checks will fail. Note that this isn't then producing the wrong
+        # gradients when backprop'ing as usual because `torch.autograd.Function.forward` has an implicit
+        # `torch.no_grad()` guard, i.e. any history doesn't affect the usual gradient computations anyway. But it does
+        # still get detected if we call `torch.autograd.grad` ourselves, which is explicitly what we don't want as per
+        # the previous paragraph.
+        assert t.is_leaf, "Internal error: please report a bug to torchsde"
+        assert y_aug.is_leaf, "Internal error: please report a bug to torchsde"
+        assert v.is_leaf, "Internal error: please report a bug to torchsde"
+
+        # This determines whether or not we will be able to backpropagate through the calling function (that is calling
+        # _unpack_y_aug). Simply, we should be able to if and only if (a) gradients are enabled and (b) the input
+        # arguments require gradient.
+        # Note that we don't fix this to True, because that implies building computational graphs, which will consume
+        # additional memory that will be unnecessary if we don't need to backpropagate.
+        requires_grad = torch.is_grad_enabled() and (t.requires_grad or y_aug.requires_grad or v.requires_grad)
+
         y, adj_y = misc.flat_to_shape(y_aug, self._shapes[:2])
-        return y.detach().requires_grad_(True), adj_y.detach().requires_grad_(True)
+
+        # To support the later differentiation wrt y, we set it to require_grad if it doesn't already.
+        if not y.requires_grad:
+            y = y.detach().requires_grad_(True)
+        return y, adj_y, requires_grad
 
     ########################################
     #                  f                   #
     ########################################
 
     def f_uncorrected(self, t, y_aug):  # For Ito additive and Stratonovich.
+        y, adj_y, requires_grad = self._unpack_y_aug(t, y_aug, v=t)  # just use t as a dummy `v`
```

Comment:
It seems a bit weird that `requires_grad` is returned by `_unpack_y_aug`, as I think the method should solely be responsible for unpacking `y_aug` as its name says. 

I think we should be able to define `requires_grad` before calling `_unpack_y_aug`. This perhaps will also read better.

Reply:
So it's there because the decision about whether to call `.detach().requires_grad_(True)` on `y` is also part of `_unpack_y_aug`'s responsibility, and it felt neatest to keep similarly-motivated pieces together.

Renaming `_unpack_y_aug` is probably simplest here.